packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,189 @@
+        const hardLimit = this.params.config.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.params.config.maxBatchSizeInBytes;
+        const softLimit = this.params.config.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.defaultAttachFlowSoftLimitInBytes;
+
+        this.attachFlowBatch = new BatchManager({
+            hardLimit,
+            softLimit,
+        });
+        this.mainBatch = new BatchManager({
+            hardLimit
+        });
+    }
+
+    public get isEmpty(): boolean {
+        return this.attachFlowBatch.length === 0 && this.mainBatch.length === 0;
+    }
+
+    public submit(message: BatchMessage) {
+        if (!this.mainBatch.push(message)) {
+            throw new GenericError(
+                "BatchTooLarge",
+                /* error */ undefined,
+                {
+                    opSize: (message.contents?.length) ?? 0,
+                    count: this.mainBatch.length,
+                    limit: this.mainBatch.options.hardLimit,
+                });
+        }
+    }
+
+    public submitAttach(message: BatchMessage) {
+        if (!this.attachFlowBatch.push(message)) {
+            // BatchManager has two limits - soft limit & hard limit. Soft limit is only engaged
+            // when queue is not empty.
+            // Flush queue & retry. Failure on retry would mean - single message is bigger than hard limit
+            this.flushInternal(this.attachFlowBatch.popBatch());
+            if (!this.attachFlowBatch.push(message)) {
+                throw new GenericError(
+                    "BatchTooLarge",
+                    /* error */ undefined,
+                    {
+                        opSize: (message.contents?.length) ?? 0,
+                        count: this.attachFlowBatch.length,
+                        limit: this.attachFlowBatch.options.hardLimit,
+                    });
+            }
+        }
+    }
+
+    public flush() {
+        this.flushInternal(this.attachFlowBatch.popBatch());
+        this.flushInternal(this.mainBatch.popBatch());
+    }
+
+    private flushInternal(rawBatch: IBatch) {
+        const processedBatch = this.maybeCompressBatch(rawBatch);
+        const clientSequenceNumber = this.sendBatch(processedBatch);
+
+        this.persistBatch(clientSequenceNumber, rawBatch.content);
+    }
+
+    private maybeCompressBatch(batch: IBatch): IBatch {
+        if (batch.content.length === 0
+            || this.params.config.compressionOptions === undefined
+            || this.params.config.compressionOptions.minimumBatchSizeInBytes >= batch.contentSizeInBytes) {
+            // Nothing to do if the batch is empty or if compression is disabled or if we don't need to compress
+            return batch;
+        }
+
+        const compressedBatch = this.params.compressor.compressBatch(batch);
+        if (compressedBatch.contentSizeInBytes < this.params.config.maxBatchSizeInBytes) {
```

Comment:
I'm confused about this check.
If this is for compressed size, then why do we reject compressed payload if it's small? I.e. if we got lucky and got 100x compression, we should use it, right?
If this is before compression (which I do not think, but spelling out thinking process), then the check should be before we attempt to compress

Reply:
Or is it the direction of comparison is wrong (i.e. should be >)?
Do we have UTs for all these interesting cases?